luigi/contrib/azureblob.py

```diff
@@ -0,0 +1,286 @@
+
+    def upload(self, tmp_path, container, blob, **kwargs):
+        logging.debug("Uploading file '{tmp_path}' to container '{container}' and blob '{blob}'".format(
+            tmp_path=self.tmp_path, container=self.container, blob=self.blob))
+        self.create_container(container)
+        lease_id = self.connection.acquire_blob_lease(container, blob)\
+            if self.exists("{container}/{blob}".format(container=container, blob=blob)) else None
+        try:
+            self.connection.create_blob_from_path(container, blob, tmp_path, lease_id=lease_id, progress_callback=kwargs.get("progress_callback"))
+        finally:
+            if lease_id is not None:
+                self.connection.release_blob_lease(container, blob, lease_id)
+
+    def download_as_bytes(self, container, blob, n=None):
+        start_range, end_range = (0, n-1) if n is not None else (None, None)
+        logging.debug("Downloading from container '{container}' and blob '{blob}' as bytes".format(
+            container=container, blob=blob))
+        return self.connection.get_blob_to_bytes(container, blob, start_range=start_range, end_range=end_range).content
+
+    def download_as_file(self, container, blob, location):
+        logging.debug("Downloading from container '{container}' and blob '{blob}' to {location}".format(
+            container=container, blob=blob, location=location))
+        return self.connection.get_blob_to_path(container, blob, location)
+
+    def create_container(self, container_name):
+        return self.connection.create_container(container_name)
+
+    def delete_container(self, container_name):
+        lease_id = self.connection.acquire_container_lease(container_name)
+        self.connection.delete_container(container_name, lease_id=lease_id)
+
+    def exists(self, path):
+        container, blob = self.splitfilepath(path)
+        return self.connection.exists(container, blob)
+
+    def remove(self, path, recursive=True, skip_trash=True):
+        container, blob = self.splitfilepath(path)
+        if not self.exists(path):
+            return False
+        lease_id = self.connection.acquire_blob_lease(container, blob)
+        self.connection.delete_blob(container, blob, lease_id=lease_id)
+        return True
+
+    def mkdir(self, path, parents=True, raise_if_exists=False):
+        container, blob = self.splitfilepath(path)
+        if raise_if_exists and self.exists(path):
+            raise FileAlreadyExists("The Azure blob path '{blob}' already exists under container '{container}'".format(
+                blob=blob, contaniner=container))
+
+    def isdir(self, path):
+        """
+        Azure Blob Storage has no concept of directories. It always returns False
+        :param str path: Path of the Azure blob storage
+        :return: False
+        """
+        return False
+
+    def move(self, path, dest):
+        return self.copy(path, dest) and self.remove(path)
```

Comment:
could `copy` succeed and `remove` fail? in that case, we have `path` in two locations

Reply:
Fixed this.